### PR TITLE
Some basic updates to perf testing. More to come.

### PR DESCRIPTION
**Pull request by brunoterkaly**
The Braveheart Team would humbly like to start the process of helping with the Engineering Playbook (EP). If all goes as planned, this is the first of many PRs to move the already excellent EP forward. We welcome input on this PR.

Looking for some input on growing this document to be as useful as possible. Here are some ideas:

#### Directions we can go

```
I.   Scenarios - What you are testing
     A. Kubernetes
        1.  Pods, Services, Network Traffic
     B. REST APIs
        1.  Testing POST, GET, PATCH, PUT
        2.  Check API response time, latency, TTFB/TTLB in various
            scenarios (in isolation and under load)
     C. Database
        1.  Query Performance
        2.  Ingestion/Export
        3.  Locks, Transactions, rollback segment
     D. Queues
II.  Performance testing frameworks
     A. Azure AzDo Pipelines
        1.  How to execute from a pipeline
     B. Performance tests as Kubernetes Jobs
III. Tools
     A. Hardware/Software to test
        1.  HTTP
        2.  Disk
        3.  Memory
        4.  Network
        5.  CPU
        6.  Network
     B. Open Source Software
        1.  Grinder, Gatling, Tsung, JMeter, Locust
IV.  Common Problems
     A. Bottlenecking — This occurs when data flow is interrupted or
        halted because there is not enough capacity to handle the workload.
     B. Poor scalability — If software cannot handle the desired number
        of concurrent tasks, results could be delayed, errors could increase,
        or other unexpected behavior could happen that affects:
        1.  Disk usage
        2.  CPU usage
        3.  Memory leaks
        4.  Operating system limitations
        5.  Poor network configuration
     C. Software configuration issues — Often settings are not set at a
        sufficient level to handle the workload.
     D. Insufficient hardware resources — Performance testing may
        reveal physical memory constraints or low-performing CPUs
V.   Performance Testing Best Practices
     A. Test Plans
     B. Involving Developers
     C. Establishing a baseline
     D. More hardware or better software
     E. Determining Performance Testing Objectives
     F. Quantifying End-User Response Time Goals
     G. Consolidating Various Types of Performance Acceptance Criteria
```

Thank you to Whitney Griffith, Tim Reilly, Michael Tarng, and Sean Teeling.

### CODE REVIEW

File: automated-testing/performance-testing/readme.md
Comment (alkofahi) [inline automated-testing/performance-testing/readme.md:79]: please re-add the link to the load-testing document

File: automated-testing/performance-testing/readme.md
Comment (alkofahi) [inline automated-testing/performance-testing/readme.md:97]: suggestion: good performance under extended periods of load

File: automated-testing/performance-testing/readme.md
Comment (alkofahi) [inline automated-testing/performance-testing/readme.md:247]: what are we going for in this section?

File: automated-testing/performance-testing/readme.md
Comment (alkofahi) [inline automated-testing/performance-testing/readme.md:253]: I think the metrics below can be explained without the table above

File: automated-testing/performance-testing/readme.md
Comment (alkofahi) [inline automated-testing/performance-testing/readme.md:285]: Since they all are one-liners, could you plz change them to bullet points or a table?

**Review by alkofahi — CHANGES_REQUESTED** [review]
Thanks, Bruno for the contribution, we have a bi-weekly meeting to discuss similar topics and how to improve upon the content, I would be glad to forward you the invite. Sharing experiences via PRs is also welcome.

File: automated-testing/performance-testing/readme.md
Comment (brunoterkaly) [reply automated-testing/performance-testing/readme.md:65]: Will follow up on all of these in the next day or so. Thanks, @alkofahi !

File: automated-testing/performance-testing/readme.md
Comment (brunoterkaly) [reply automated-testing/performance-testing/readme.md:79]: Yes, agreed. Updated to avoid names.

File: automated-testing/performance-testing/readme.md
Comment (brunoterkaly) [reply automated-testing/performance-testing/readme.md:97]: Agreed. More clear. Easier to read.

File: automated-testing/performance-testing/readme.md
Comment (brunoterkaly) [reply automated-testing/performance-testing/readme.md:247]: I'll try to explain. the idea I could explain better is that measuring performance is dependent on the capacity of the network type.

File: automated-testing/performance-testing/readme.md
Comment (brunoterkaly) [reply automated-testing/performance-testing/readme.md:253]: Agreed. It made sense to remove it. Put in some basic troubleshooting steps instead.

File: automated-testing/performance-testing/readme.md
Comment (brunoterkaly) [reply automated-testing/performance-testing/readme.md:285]: Good idea. They now all look good as tables.

File: automated-testing/performance-testing/readme.md
Comment (alkofahi) [reply automated-testing/performance-testing/readme.md:79]: The title used to link to "./load-testing.md" that focuses on load-testing. Can you re-add the link, either to the title or somewhere in the paragraph.

File: automated-testing/performance-testing/readme.md
Comment (alkofahi) [reply automated-testing/performance-testing/readme.md:97]: Sorry, I think I did not elaborate. I meant to change the text in the paragraph "can maintain
good performance when extended periods of load." to "can maintain
good performance under extended periods of load."...I think the title should remain "Endurance testing" as that is the commonly known name for this kind of test

**Review by alkofahi — CHANGES_REQUESTED** [review]

**Comment by fnocera** [comment]
Once #607 is merged in all outstanding PRs including this one will have to make a few changes and merge master to adopt the new MKDocs structure. This main change is that all the md files move into a `docs` folder. Please don't hesitate to reach out with any questions and a number of us are more than happy to help with this process - reach out to @fnocera @TessFerrandez @omri374 and @brockneedscoffee

**Review by alkofahi — APPROVED** [review]

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:82]:
```suggestion
performance characteristics of a system, when the system faces the load volumes
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:83]:
```suggestion
expected during production operation. An **Endurance Test** or a **Soak
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:89]:
```suggestion
This is the subcategory of performance testing that focuses on validating the
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:81]:
```suggestion
This is the subcategory of performance testing that focuses on validating the
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:115]:
```suggestion
When executing the various types of tests, whether it is stress,
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:116]:
```suggestion
endurance, spike, or chaos testing, it is important to capture various
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:130]:
```suggestion
The classic example of how performance in one area can affect performance in
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:152]:
```suggestion
| % Idle Time | This is a measure of the percentage of time that the disk was idle. i.e. there are no pending disk requests from the operating system waiting to be completed. A low number here is a positive sign that disk has excess capacity to service or write requests from the operating system.  |
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:241]:
```suggestion
Performance testing activities vary depending on subcategory of performance
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:243]:
```suggestion
follow the link to the subcategory of performance tests listed above. Following
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:244]:
```suggestion
are some activities that may be involved:
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:248]:
```suggestion
This will generally include identifying and defining the goals and constrains
```
```suggestion
This will generally include identifying and defining the goals and constraints
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:255]:
```suggestion
- Defining the load the application should be tested with
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:259]:
```suggestion
- Establish what tools will be used for the tests
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:272]:
```suggestion
  previous section around load testing, chaos testing, stress testing, and spike testing
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:276]:
```suggestion
- Collect the test results
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:278]:
```suggestion
- Analyze the test results
```

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:280]: Follow the same pattern as in the earlier sections with active sentences. i.e. Depending on the scenario, modify the application or configuration and repeat the tests.

**Review by TessFerrandez — APPROVED** [review]
I like the content, but suggest some language cleanup to help the reader understand the content. Approving with suggestions

File: automated-testing/performance-testing/readme.md
Comment (TessFerrandez) [inline automated-testing/performance-testing/readme.md:4]:
```suggestion
Performance Testing is an overloaded term that is used to refer to several subcategories of performance related testing, each of which has different purpose.
```

**Comment by fnocera** [comment]
@brunoterkaly could you resolve conflicts and then we can get this merged in? Thanks!

File: automated-testing/performance-testing/readme.md
Comment (alkofahi) [reply automated-testing/performance-testing/readme.md:13]: this seems to re-iterate the quote above. I removed it!

**Comment by alkofahi** [comment]
I revised the doc based on PR feedback, but was not able to update the PR with latest changes even after the merge! I am closing this one in favor of this PR https://github.com/microsoft/code-with-engineering-playbook/pull/667

**Pull request closed without merging**
